### ISSUE TITLE
[Epic] Active replication

### ISSUE THREAD

**Issue by frankmcsherry** [MaterializeInc/materialize#11240]
https://github.com/MaterializeInc/materialize/pull/11335 added support for active replication, relying  on the *determinism* and *idempotence* of the COMPUTE layer. Having two (or more) copies of the instance up and running *should* result in reconcilable results.

### Goals
* Allow manually adding/removing replicas for clusters

### Non-Goals
* Auto-replicating clusters 
* Support for sinks (blocked by https://github.com/MaterializeInc/materialize/issues/11501)

### Acceptance criteria

Users can launch replicated clusters. Materialize will serve requests using all replicas of a cluster. Errors within one replica are transparent while sufficient replicas are online, which means that the user will not observe a functional degradation, but can observe a qualitative degradation of Materialize's service.

Adding or removing replicas does not cause downtime. It can take a moment before a new replica completes initial data loading and can serve requests.

### Subtasks

* [x] [design/20220413_cluster_replica.md](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/20220413_cluster_replica.md)
  * [x] #11808 
  * [x] #11809
  * [x] #11810
  * [x] MaterializeInc/cloud#2753
  * [x] MaterializeInc/cloud#2754 (see also: [Slack](https://materializeinc.slack.com/archives/CL68GT3AT/p1649854802804479))
* [x] https://github.com/MaterializeInc/materialize/issues/12959 to enable reporting the health of a replica outwards
* [x] https://github.com/MaterializeInc/materialize/issues/12520
* [x] https://github.com/MaterializeInc/materialize/issues/12816

### Testing tasks
* [x] Add a functional test with targeted failure injection (kills, suspends, suspends + resumes) on individual replicas to confirm that the cluster remains responsive throughout
* ~[ ] A Nightly CI job that runs testdrive with 4 replicas ( increase from the current value of 2 replicas)~
* ~[ ] Add a feature benchmark that runs a multi-replica scenario (e.g. 10 replicas)~
* ~[ ] Add functionality to Zippy to allow it to create, drop and insert failures in replicas~

### Related

* [design/20220413_cluster_replica.md](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/20220413_cluster_replica.md)
* https://github.com/MaterializeInc/materialize/issues/11782 (moved to M2)
* https://github.com/MaterializeInc/materialize/issues/12808 (moved to M2)

### Open Questions

* What needs to be done to revive instances that have stalled out? How is this communicated through the layers (e.g. from an otherwise local deduplicator up through the orchestration layer).
* One can only do active replication for *definite* collections; if the results could diverge, it is not clear on what basis we can return correct answers.

**Comment by teskje** [comment]
I did some manual end-to-end tests on a personal cloud stack, to verify everything we currently have works as intended.

# Test 1: Replica Configuration

This test verified that the replica configuration parameters for size and AZ are respected by our orchestrator.

## Setup

```sql
CREATE CLUSTER replica_test
  REPLICA a (SIZE 'xsmall', AVAILABILITY ZONE 'use1-az2'),
  REPLICA b (SIZE 'small', AVAILABILITY ZONE 'use1-az4'),
  REPLICA c (SIZE 'medium', AVAILABILITY ZONE 'use1-az6');
```

## Observations

The above cluster creation command spawns 7 pods in 3 StatefulSets:

* `compute-cluster-2-replica-2-0` in us-east-1a (= use1-az2)
* `compute-cluster-2-replica-3-{0,1}` in us-east-1b (= use1-az4)
* `compute-cluster-2-replica-4-{0,1,2,3}` in us-east-1c (= use1-az6)

The number of pods per replica matches the `scale` value associated with the respective replica sizes. CPU and memory container limits match as well.

# Test 2: Crashing Replicas

This test verified that Materialize continues to serve requests in the face of replicas crashing.

## Setup

```sql
CREATE CLUSTER replica_test
  REPLICA a (SIZE 'xsmall'),
  REPLICA b (SIZE 'xsmall'),
  REPLICA c (SIZE 'xsmall');

SET cluster = replica_test;

CREATE SOURCE order_source FROM PUBNUB
 SUBSCRIBE KEY 'sub-c-4377ab04-f100-11e3-bffd-02ee2ddab7fe'
 CHANNEL 'pubnub-market-orders';

CREATE MATERIALIZED VIEW orders AS
SELECT
    (text::jsonb)->>'bid_price' AS bid_price,
    (text::jsonb)->>'order_quantity' AS order_quantity,
    (text::jsonb)->>'symbol' AS symbol,
    (text::jsonb)->>'trade_type' AS trade_type,
    to_timestamp(((text::jsonb)->'timestamp')::bigint) AS timestamp
FROM order_source;

COPY (TAIL (SELECT * FROM orders)) TO STDOUT;
```

* Crashing pods were simulated by killing them in k9s.
* Crashing nodes were simulated by terminating them in the EC2 console.

## Observations

* Crashing one or two of the replica pods doesn't affect the TAIL execution (from a user's POV).
* Crashing one or two of the replica nodes doesn't affect the TAIL execution either.
* If all three replica pods (or replica nodes) crash, the TAIL hangs.
  * Eventually, K8S restarts the replicas and the TAIL continues.

# Test 3: Stalling Replicas

This test verified that Materialize continues to serve requests in the face of replicas slowing down significantly.

## Setup

```
(same as Test 2)
```

* Stalling replicas was simulated by reducing their container CPU limit from "1600m" (1.6 CPUs) to "1m" (0.001 CPUs).

## Observations

* Stalling one or two of the replicas doesn't affect the TAIL execution.
* If all three replicas are stalled, the TAIL hangs.
  * Un-stalling one replica also un-hangs the TAIL.
